clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -775,8 +776,6 @@ public void handle(SyncGroupResponse syncResponse,
                     }
                 }
             } else {
-                requestRejoin();
```

Comment:
We can remove this since it is a bit redundant now as we call for each case if necessary.

Reply:
I meant the latter: we call that inside the conditions already -- for those fatal errors, we do not need to call this anyways since the consumer will throw and crash.

Reply:
@guozhangwang I think something may have been messed up during a merge/rebase: I no longer see `requestRejoinOnResponseError` being invoked anywhere

Reply:
I added that function for sync group handler that handles retriable `COORDINATOR_NOT_AVAILABLE / NOT_COORDINATOR` and any unexpected error. After the refactoring PR they are not all fall into the `joinGroupIfNeeded` in

```
final RuntimeException exception = future.exception();

                resetJoinGroupFuture();

                if (exception instanceof UnknownMemberIdException ||
                    exception instanceof IllegalGenerationException ||
                    exception instanceof RebalanceInProgressException ||
                    exception instanceof MemberIdRequiredException)
                    continue;
                else if (!future.isRetriable())
                    throw exception;

                resetStateAndRejoin(String.format("rebalance failed with retriable error %s", exception));
                timer.sleep(rebalanceConfig.retryBackoffMs);
```

This is part of the principle I mentioned:

```
We may reset generationa and request rejoin in two different places: 1) in join/sync-group handler, and 2) in joinGroupIfNeeded, when the future is received. The principle is that these two should not overlap, and 2) is used as a fallback for those common errors from join/sync that we do not handle specifically.
```

But I forgot to remove this function as part of the second pass; will remove.

Reply:
Ok cool, thanks. One last question then: after this refactoring, since we no longer call `requestRejoinOnResponseError` below, should we re-add the `requestRejoin()` call here? Or add a `requestRejoin` to the specific cases in the SyncGroup handler, eg
```
} else if (error == Errors.REBALANCE_IN_PROGRESS) {
    log.info("SyncGroup failed: The group began another rebalance. Need to re-join the group. " +
                      "Sent generation was {}", sentGeneration);
    future.raise(error);
}
```

Reply:
I think we do not need to, since it would be called on `resetStateAndRejoin(String.format("rebalance failed with retriable error %s", exception));` --- previously we are calling rejoin double times.

Reply:
Hmm...but `resetStateAndRejoin(String.format("rebalance failed with retriable error %s", exception));` is only called in `joinGroupIfNeeded`  which is only called in `ensureActiveGroup`, which is in turn only invoked in `ConsumerCoordinator#poll`.
That said,  inside `SyncGroupResponseHandler#handle` we would already have `rejoinNeeded = true` and only set it to false if the SyncGroup succeeds. So for that reason I guess we don't need the `requestRejoin` anywhere inside the SyncGroup handler